test/parallel/test-http-upgrade-server-with-body.mjs

```diff
@@ -0,0 +1,209 @@
+import * as common from '../common/index.mjs';
+import * as assert from 'assert';
+import * as http from 'http';
+import * as net from 'net';
+
+async function createServer({ expectedBodyLength }) {
+  const server = http.createServer();
+
+  server.on('request', common.mustNotCall());
+
+  server.on('upgrade', function(req, socket, upgradeHead) {
+    // Confirm the upgrade:
+    socket.write('HTTP/1.1 101 Switching Protocols\r\n' +
+                'Upgrade: custom-protocol\r\n' +
+                'Connection: Upgrade\r\n' +
+                '\r\n');
+
+    // Read and validate the request body, if and only if expectedBodyLength is set:
+    if (expectedBodyLength !== undefined) {
+      let reqBody = '';
+      req.on('data', (str) => { reqBody += str; });
+      req.on('end', common.mustCall(() => {
+        assert.strictEqual(reqBody.length, expectedBodyLength);
+      }));
+    }
+
+    let socketData = upgradeHead;
+
+    // Collect the raw protocol data and check it:
+    socket.on('data', common.mustCall((d) => {
+      socketData = Buffer.concat([socketData, d]);
+    }));
+
+    socket.on('end', common.mustCall(() => {
+      assert.strictEqual(socketData.toString(), 'upgrade head\npost-upgrade message');
+      socket.end();
+    }));
+  });
+
+  await new Promise((resolve) => server.listen(0, () => resolve()));
+
+  return server;
+}
+
+async function testUpgradeWithBody() {
+  const server = await createServer({ expectedBodyLength: 12 });
+
+  const conn = net.createConnection(server.address().port);
+  conn.setEncoding('utf8');
+
+  await new Promise((resolve) => conn.on('connect', resolve));
+
+  // Write request headers, body & upgrade head all together:
+  conn.write(
+    'POST / HTTP/1.1\r\n' +
+    'host: localhost\r\n' +
+    'Upgrade: custom-protocol\r\n' +
+    'Connection: Upgrade\r\n' +
+    'transfer-encoding: chunked\r\n' +
+    '\r\n' +
+    'C\r\nrequest body\r\n' + // 12 byte body sent immediately
+    '0\r\n\r\n' +
+    'upgrade head'
+  );
+
+  const response = await new Promise((resolve) => conn.once('data', resolve));
+  assert.ok(response.startsWith('HTTP/1.1 101 Switching Protocols\r\n'));
```

Comment:
There is no guarantee that this data is received in a single chunk.

Reply:
True, although I have run these in a tight loop with & without load for thousands of runs without that ever happening. Tests now refactored substantially, I think they don't depend on specific chunking behaviour but they do try to encourage it to encourage test coverage of awkward chunking scenarios.